src/app/clusters/service-area-server/service-area-server.cpp

```diff
@@ -0,0 +1,1273 @@
+    ret_value = true;
+
+
+exit:
+    return ret_value;
+}
+
+
+//*************************************************************************
+// Progress list manipulators
+
+bool Instance::AddPendingProgressElement(uint32_t aLocationId)
+{
+    bool     ret_value = false;
+    uint32_t dummyIndex;
+
+    // create progress element
+    Structs::ProgressStruct::Type inactiveProgress = { aLocationId, OperationalStatusEnum::kPending};
+
+    // check max# of list entries
+    VerifyOrExit((mDelegate->GetNumberOfProgressElements() < kMaxNumProgressElements),  // Note: progress elements must uniquely map to supported locations, so max size is the same
+                ChipLogError(Zcl,  "AddPendingProgressElement - maximum number of entries"));
+
+    // For each entry in this list, the LocationID field SHALL match an entry on the SupportedLocations attribute's list.
+    VerifyOrExit(IsSupportedLocation(aLocationId),
+                ChipLogError(Zcl,  "AddPendingProgressElement - not a supported location %u", aLocationId));
+
+    // Each entry in this list SHALL have a unique value for the LocationID field.
+
+    VerifyOrExit(!IsProgressElement(aLocationId),
+                ChipLogError(Zcl,  "AddPendingProgressElement - progress element already exists for location %u", aLocationId));         
+
+
+    VerifyOrExit(mDelegate->AddProgressElement(inactiveProgress, dummyIndex), /* if false, should be logged as error in delegate function */);
+
+
+    // success
+    ret_value = true;
+    NotifyProgressChanged();
+
+exit:
+    return ret_value;
+}
+
+
+bool Instance::SetProgressStatus(uint32_t aLocationId, OperationalStatusEnum opStatus)
+{
+    bool     ret_value = false;
+    uint32_t listIndex;
+
+    Structs::ProgressStruct::Type progressElement;
+
+    VerifyOrExit(mDelegate->GetProgressElementById(aLocationId, listIndex, progressElement),
+                ChipLogError(Zcl,  "SetProgressStatus - progress element does not exist for location %u", aLocationId));
+
+
+    // if status value not changing, no need to modify the existing element
+    VerifyOrExit((progressElement.status != opStatus), ret_value = true);
+
+    // set the progress status in the local copy
+    progressElement.status = opStatus;
+
+    // TotalOperationalTime SHALL be null if the Status field is not set to Completed or Skipped.
+    if ((opStatus != OperationalStatusEnum::kCompleted) &&
+        (opStatus != OperationalStatusEnum::kSkipped))
+    {
+        progressElement.totalOperationalTime.Value().SetNull();
+    }
+
+    // add the updated element to the progress attribute
+    VerifyOrExit(mDelegate->ModifyProgressElement(listIndex, progressElement), /* if false, should be logged as error in delegate function */);
+
+
+    // success
+    ret_value = true;
+    NotifyProgressChanged();
+
+ exit:
+    return ret_value;
+}
+
+
+bool Instance::SetProgressTotalOperationalTime(uint32_t aLocationId, const DataModel::Nullable<uint32_t> & aTotalOperationalTime)
```

Comment:
Should we combine this with the `SetProgressStatus` method by having the default `aTotalOperationTime` set to null?

Reply:
It is fine as it is.